ext/opcache/jit/zend_jit_arm64.dasc

```diff
@@ -0,0 +1,15160 @@
+#define LDRB_STRB_PIMM  MAX_IMM12      // ldrb/strb insn
+
+#define B_IMM26         (1<<27)        // signed imm26 * 4
+
+static bool arm64_may_use_b(const void *addr)
+{
+	if (addr >= dasm_buf && addr < dasm_end) {
+		return (((char*)dasm_end - (char*)dasm_buf) < B_IMM26);
+	} else if (addr >= dasm_end) {
+		return (((char*)addr - (char*)dasm_buf) < B_IMM26);
+	} else if (addr < dasm_buf) {
+		return (((char*)dasm_end - (char*)addr) < B_IMM26);
+	}
+	return 0;
+}
+
+#include "Zend/zend_cpuinfo.h"
+
+#ifdef HAVE_VALGRIND
+# include <valgrind/valgrind.h>
+#endif
+
+/* The generated code may contain tautological comparisons, ignore them. */
+#if defined(__clang__)
+# pragma clang diagnostic push
+# pragma clang diagnostic ignored "-Wtautological-compare"
+# pragma clang diagnostic ignored "-Wstring-compare"
+#endif
+
+const char* zend_reg_name[] = {
+	"x0", "x1", "x2", "x3", "x4", "x5", "x6", "x7",
+	"x8", "x9", "x10", "x11", "x12", "x13", "x14", "x15",
+	"x16", "x17", "x18", "x19", "x20", "x21", "x22", "x23",
+	"x24", "x25", "x26", "x27", "x28", "x29", "x30", "sp",
+	"v0", "v1", "v2", "v3", "v4", "v5", "v6", "v7",
+	"v8", "v9", "v10", "v11", "v12", "v13", "v14", "v15",
+	"v16", "v17", "v18", "v19", "v20", "v21", "v22", "v23",
+	"v24", "v25", "v26", "v27", "v28", "v29", "v30", "v31"
+};
+
+#ifdef HAVE_GCC_GLOBAL_REGS
+# define GCC_GLOBAL_REGS 1
+#else
+# define GCC_GLOBAL_REGS 0
+#endif
+
+# define ZREG_FCARG1x ZREG_X0
+# define ZREG_FCARG2x ZREG_X1
+
+#if ZTS
+static size_t tsrm_ls_cache_tcb_offset = 0;
+#endif
+
+/* By default avoid JITing inline handlers if it does not seem profitable due to lack of
+ * type information. Disabling this option allows testing some JIT handlers in the
+ * presence of try/catch blocks, which prevent SSA construction. */
+#ifndef PROFITABILITY_CHECKS
+# define PROFITABILITY_CHECKS 1
+#endif
+
+|.type EX, zend_execute_data, FP
+|.type OP, zend_op
+|.type ZVAL, zval
+
+|.actionlist dasm_actions
+
+|.globals zend_lb
+static void* dasm_labels[zend_lb_MAX];
+
+|.section code, cold_code, jmp_table
+
+#define IS_SIGNED_32BIT(val) ((((intptr_t)(val)) <= 0x7fffffff) && (((intptr_t)(val)) >= (-2147483647 - 1)))
+
+#define BP_JIT_IS 6
+
+/* helper: determine whether an immediate value can be encoded as the immediate operand of logical instructions */
+static int logical_immediate_p (uint64_t value, uint32_t reg_size)
+{
+	/* fast path: power of two */
+	if (value > 0 && !(value & (value - 1))) {
+		return 1;
+	}
+
+	// TODO: slow path
+	return 0;
+}
+
+/* Not Implemented Yet */
+|.macro NIY
+||	//ZEND_ASSERT(0);
+|	brk #0
+|.endmacro
+
+|.macro NIY_STUB
+||	//ZEND_ASSERT(0);
+|	brk #0
+|.endmacro
+
+|.macro ADD_HYBRID_SPAD
+||#ifndef ZEND_VM_HYBRID_JIT_RED_ZONE_SIZE
+|	add sp, sp, HYBRID_SPAD
+||#endif
+|.endmacro
+
+|.macro SUB_HYBRID_SPAD
+||#ifndef ZEND_VM_HYBRID_JIT_RED_ZONE_SIZE
+|	sub sp, sp, HYBRID_SPAD
+||#endif
+|.endmacro
+
+|.macro LOAD_ADDR, reg, addr
+|	// 48-bit virtual address
+||	if (((uintptr_t)(addr)) == 0) {
```

Comment:
@shqking what do you suggest?

Reply:
Thanks for spotting that. Yes. I agree that one assertion should be added here. I will update it soon.
52-bit virtual address is supported in newer ARMv8.2-LVA (See https://opensource.com/article/20/12/52-bit-arm64-kernel). However to the best of my knowledge, there are few such chips in the market.
Java JDK currently only supports  48-bit virtual address, and I suppose putting an assertion might be one good option for now and we may want to support 52-bit address when needed in the future.

Reply:
But I don't think it would be a good solution to introduce a new variable inside the macro, i.e. `uintptr_t addr = (uintptr_t) ptr;`, which would lead to redefinition of `addr` for consecutive uses of macro `LOAD_ADDR`. 

From my view, it's a problem to conduct the type conversion inside the definition of macro, or at all the callers of this macro. @dstogov May I have your opinion here?

Reply:
@shqking I afraid, assertion may conflict with "address tagging technology" (the high byte of the address may be ignored).

Introducing new variable would require wrapping macro into ``do { ... } while (0)`` block, or even C function. I prefer to think about this later.

Reply:
Yes. Agree. The higher bits might be used by some security features, e.g., MTE, PAC.